torchgen/model.py

```diff
@@ -946,6 +968,38 @@ def native_function_class_name(self) -> Optional[str]:
             # updating every kernel definition + callsite of every in-tree aten kernel.
             return None
 
+    def should_gen_dispatchless_composite(self, f: NativeFunction) -> bool:
+        if f.func.name in self.index:
+            return False
+        return (
+            len(f.composite) > 0
+            and self.dispatch_key in COMPOSITE_DISPATCH_KEYS
+        )
+
+    def dispatchless_composite_struct(self, f: NativeFunction) -> str:
+        name = f.func.name.unambiguous_name()
+        return f"{name}_{self.dispatch_key.lower()}"
+
+    def dispatchless_composite_kernel(self, f: NativeFunction) -> str:
+        struct_name = self.dispatchless_composite_struct(f)
+
+        # Similar to cpp.name().
+        # Couldn't use it due to circular dependency issues.
+        name = str(f.func.name.name)
+        if f.func.is_out_fn():
+            name += "_out"
+
+        return f"{name}<{struct_name}>"
+
+    def has_registered_kernel(
+            self, f: NativeFunction, g: Optional[NativeFunctionsGroup]
+    ) -> bool:
+        if (
+                g is not None
+                and g.structured
+        ):
+            return self.has_kernel(g) or self.dispatch_key == DispatchKey.Meta
+        return self.has_kernel(f)
```

Comment:
Is there a better way to do this? While this seems to work, it feels a bit hacky...